src/commissaire_service/clusterexec/__init__.py

```diff
@@ -0,0 +1,305 @@
+# Copyright (C) 2016  Red Hat, Inc
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import datetime
+
+from commissaire.models import (
+    ClusterDeploy, ClusterUpgrade, ClusterRestart, Cluster, Hosts)
+from commissaire.util.ssh import TemporarySSHKey
+
+from commissaire_service.oscmd import get_oscmd
+from commissaire_service.service import CommissaireService
+from commissaire_service.transport import ansibleapi
+
+
+class ClusterExecService(CommissaireService):
+    """
+    Executes operations over a cluster by way of remote shell commands.
+    """
+
+    def __init__(self, exchange_name, connection_url):
+        """
+        Creates a new ClusterExecService.
+
+        :param exchange_name: Name of the topic exchange
+        :type exchange_name: str
+        :param connection_url: Kombu connection URL
+        :type connection_url: str
+        """
+        queue_kwargs = [
+            {'routing_key': 'jobs.clusterexec'}
+        ]
+        super().__init__(exchange_name, connection_url, queue_kwargs)
+
+    def _execute(self, model_instance, command_args, finished_hosts_key):
```

Comment:
This function feels too long. It would be ideal to break down each test into a separate function, to promote re-use, but also enhance the readability and intent of each check.